packages/type-utils/src/TypeOrValueSpecifier.ts

```diff
@@ -0,0 +1,202 @@
+            },
+          ],
+        },
+        source: {
+          type: 'string',
+        },
+      },
+      required: ['from', 'name'],
+    },
+    {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        from: {
+          type: 'string',
+          const: 'lib',
+        },
+        name: {
+          oneOf: [
+            {
+              type: 'string',
+            },
+            {
+              type: 'array',
+              minItems: 1,
+              uniqueItems: true,
+              items: {
+                type: 'string',
+              },
+            },
+          ],
+        },
+      },
+      required: ['from', 'name'],
+    },
+    {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        from: {
+          type: 'string',
+          const: 'package',
+        },
+        name: {
+          oneOf: [
+            {
+              type: 'string',
+            },
+            {
+              type: 'array',
+              minItems: 1,
+              uniqueItems: true,
+              items: {
+                type: 'string',
+              },
+            },
+          ],
+        },
+        source: {
+          type: 'string',
+        },
+      },
+      required: ['from', 'name', 'source'],
+    },
+    {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        from: {
+          type: 'array',
+          minItems: 1,
+          uniqueItems: true,
+          items: {
+            type: 'string',
+            enum: ['file', 'lib', 'package'],
+          },
+        },
+        name: {
+          type: 'string',
+        },
+      },
+      required: ['from', 'name'],
+    },
+  ],
+};
+
+function specifierNameMatches(
+  type: ts.Type,
+  name: string | Array<string>,
+): boolean {
+  if (typeof name === 'string') {
+    name = [name];
+  }
+  return name.some(item => item === type.getSymbol()?.escapedName);
+}
+
+function typeMatchesFileSpecifier(
+  specifier: FileSpecifier,
+  declarationFiles: Array<ts.SourceFile>,
+  program: ts.Program,
+): boolean {
+  if (
+    !declarationFiles.some(declaration =>
+      declaration.fileName.startsWith(program.getCurrentDirectory()),
+    )
+  ) {
+    return false;
+  }
+  const source = specifier.source;
+  if (source === undefined) {
+    return true;
+  }
+  const specifierPath = path.join(program.getCurrentDirectory(), source);
+  return declarationFiles.some(
+    declaration => declaration.fileName === specifierPath,
+  );
+}
+
+export function typeMatchesSpecifier(
+  type: ts.Type,
+  specifier: TypeOrValueSpecifier,
+  program: ts.Program,
+): boolean {
+  if (typeof specifier === 'string') {
+    return specifierNameMatches(type, specifier);
+  }
+  if (!specifierNameMatches(type, specifier.name)) {
+    return false;
+  }
+  const declarationFiles =
+    type
+      .getSymbol()
+      ?.getDeclarations()
+      ?.map(declaration => declaration.getSourceFile()) ?? [];
```

Comment:
[Praise] Kinda nifty how this line is used in all three cases. I find it very clean+neat. And in general this is a pretty succinct set of functions. Nice 😄

Reply:
Thanks, I spent quite some time on making this be nice :D Keep in mind that this set of functions only works for a type, not a value (in the RFC you talked about this specifier format also being used for values, I can't quite think of an example, but that would need a separate function.)